## CSE 5243 Intro To Data Mining Final Project: Predicting NFL Outcomes

# Intro

### The objective of this project will be attempting to make a decent model at predicting the outcome of NFL games and comparing them to Vegas' models for gambling. The model will calculate an expected point differential and compare it with the Vegas spread, determining whether betting on the home team or the away team will "cover" the spread. Considering the complexity of the NFL and how many different moving pieces and variables exist, I am honestly not sure how well this model will perform. I intend to submit the completed model whether the results are good and bad and probably further tweak it in the future.

# Data Pre-Processing

In [2]:
import numpy as np
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)
pd.set_option('display.max_rows', None)
np.set_printoptions(suppress=True, precision=4)
import math
from sklearn import preprocessing

In [3]:
games = pd.read_csv('gameData.csv')

In [4]:
games[['wins_home','losses_home','wins_away','losses_away']] = preprocessing.MinMaxScaler().fit_transform(games[['wins_home','losses_home','wins_away','losses_away']])

# Principal Component Analysis

In [5]:
gamesDummied = pd.get_dummies(games).reset_index(drop=True)
gamesDummied.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5324 entries, 0 to 5323
Columns: 14481 entries, losses_away to team_home_Washington Redskins
dtypes: float64(5), int64(4), uint8(14472)
memory usage: 73.8 MB


First, in order to handle the categorical variables, all column values will be "dummied". Unfortunately, with every player and team name being a unique value for both home and away columns, this translates to our datasets containing 14k+ features. To save on computation time if this were to run on a weekly basis, we will be using PCA to limit our features to a more managable number without losing too much varianace.

In [6]:
from sklearn.decomposition import PCA
pca = PCA(n_components=0.99999, svd_solver='full')
outputs = gamesDummied[['vegas_expected_home_win_margin','score_home','score_away','schedule_season','home_bet']]
gameComponents = pca.fit_transform(gamesDummied.drop(columns=['vegas_expected_home_win_margin','score_home','score_away','home_bet','schedule_season']))

In [7]:
print(np.sum(pca.explained_variance_ratio_))
pcaGames = pd.DataFrame(data = gameComponents, columns = map(lambda x: 'princomp' + str(x), range(0,len(gameComponents[0]))))

0.9999917212007765


I've arbitrarily decided that a threshold of 99.999% retained variance would be acceptable for the principal component analysis. 

# Regression Models (Support Vector Machine and Multilayer Perceptron Model)

In [10]:
pcaGames[['vegas_expected_home_win_margin','score_home','score_away','schedule_season', 'home_bet']] = outputs[['vegas_expected_home_win_margin','score_home','score_away','schedule_season','home_bet']]

train = pcaGames[pcaGames.schedule_season <= 2018]
test = pcaGames[pcaGames.schedule_season >= 2019]

Since the point of the model will be to predict the outcomes of future games, the two most recent seasons are used as a test set while all previous seasons since 2000 are used as a training set. This makes it so that our data is no longer random, but it would not make sense to have future games predict those in the past.

## Poisson Regressor

### More Data Prep

For the SVM regressor, the 'schedule_season' column is first used to create a weight value for each datapoint. More recent games are given higher weights so that they are more relevant when training the dataset. This is necessary since the performance of older varieties of teams should have little bearing on more recent games. A customized exponential function is used here to weight games in a mostly arbitrary manner; the least recent season is given a weight of '1' while the most recent is given a weight of about '5'. The range of weights was considered to be higher, but since some of this is already accounted for by including team rosters for each season (which will be more similar to other nearby seasons), this final weight function was decided on.

In [17]:
trainX = train.drop(columns=['vegas_expected_home_win_margin','score_home','score_away','home_bet','schedule_season'])
trainYHome = train['score_home']
trainYAway = train['score_away']

Our train input features will end up being all of the dummied variables (players, head coaches, team names, stadium locations) as well as the normalized wins and losses each team has coming into a game. The output feature will be the point differential in reference to the home team. The vegas spread is dropped since it is essentially a regression prediction in and of itself, with the goal of the project being attempting to 'out-predict' it. The 'schedule_season' is represented in the sample weights and the 'class' will be used to judge our betting classifier, so these too are dropped.

### Hypertuning

In [54]:
from sklearn.linear_model import PoissonRegressor
from sklearn.metrics import mean_squared_error

testX = test.drop(columns=['vegas_expected_home_win_margin','score_home','score_away','home_bet','schedule_season'])
testYHome = test['score_home']
testYAway = test['score_away']

clfHome = PoissonRegressor(alpha=alpha,max_iter=iteration)
clfHome.fit(trainX, trainYHome)

predYHome = clfHome.predict(testX)
print('Home score MSE is',mean_squared_error(testYHome,predYHome))

clfAway = PoissonRegressor(alpha=alpha,max_iter=iteration)
clfAway.fit(trainX, trainYAway)

predYAway = clfAway.predict(testX)
print('Home score MSE is',mean_squared_error(testYAway,predYAway))

Home score MSE is 95.16670109079493
Home score MSE is 101.53970141552658


In [50]:
predictions = pd.DataFrame(data=np.column_stack((predYHome,predYAway)),columns=['Home','Away'])
predictions['home_point_lead'] = predictions['Home'] - predictions['Away']
predictions['vegas_spread'] = test['vegas_expected_home_win_margin'].to_numpy()

def calculateHomeBet(row):
    if (row['home_point_lead'] > row['vegas_spread']):
        row['home_bet'] = 1
    else:
        row['home_bet'] = 0
    return row

predictions = predictions.apply(calculateHomeBet, axis=1)
predictions

Home   Away  home_point_lead  vegas_spread  home_bet
0   23.665 20.360            3.306         3.000     1.000
1   20.910 20.812            0.097        -3.000     1.000
2   21.836 22.318           -0.482        -2.000     1.000
3   20.790 19.518            1.272         5.500     0.000
4   23.832 21.437            2.395         7.000     0.000
5   22.595 23.639           -1.044        -3.500     1.000
6   23.755 20.539            3.216        -6.000     1.000
7   20.448 21.044           -0.596        -7.000     1.000
8   23.685 19.569            4.115         3.500     1.000
9   25.538 19.001            6.537         5.500     1.000
10  23.250 22.197            1.052         2.500     0.000
11  24.740 17.903            6.837        10.000     0.000
12  25.048 19.808            5.240         9.000     0.000
13  22.711 20.321            2.390         1.000     1.000
14  27.671 21.746            5.926         6.500     0.000
15  22.875 22.515            0.360        -3.000     1.000
16  23.835 20.414            3.421         6.500     0.000
17  23.404 24.377           -0.973        -1.000     1.000
18  25.397 18.092            7.305        13.000     0.000
19  23.244 22.466            0.778         1.000     0.000
20  23.464 19.189            4.275        -3.000     1.000
21  21.762 23.516           -1.754        -1.000     0.000
22  24.015 20.719            3.296         3.000     1.000
23  23.445 19.477            3.968         7.000     0.000
24  21.737 24.415           -2.677        -2.000     0.000
25  20.562 23.249           -2.687       -18.000     1.000
26  22.425 20.410            2.015        -1.000     1.000
27  24.490 26.304           -1.814        -7.000     1.000
28  24.542 21.050            3.492         4.000     0.000
29  23.274 20.912            2.362         3.500     0.000
30  19.919 21.055           -1.136        -6.000     1.000
31  24.092 22.797            1.295        -6.500     1.000
32  21.257 18.554            2.702        -2.000     1.000
33  21.356 20.322            1.034         2.500     0.000
34  22.653 20.677            1.976         6.000     0.000
35  20.406 22.911           -2.505        -4.000     1.000
36  24.355 19.889            4.467        21.500     0.000
37  24.240 20.411            3.829         7.000     0.000
38  24.288 21.254            3.034         1.500     1.000
39  22.686 20.085            2.601         5.000     0.000
40  23.960 20.451            3.509        -3.000     1.000
41  23.510 18.154            5.357         9.000     0.000
42  27.531 17.895            9.636        21.000     0.000
43  24.012 18.727            5.285         4.000     1.000
44  21.609 21.185            0.423         6.000     0.000
45  23.345 22.801            0.544         5.000     0.000
46  21.671 20.508            1.162         5.000     0.000
47  20.675 21.249           -0.574        -5.000     1.000
48  24.464 22.597            1.867         3.500     0.000
49  20.089 20.808           -0.718        -5.500     1.000
50  23.427 20.107            3.321         3.500     0.000
51  24.989 18.790            6.199         7.000     0.000
52  20.794 22.934           -2.140        -7.000     1.000
53  22.222 19.972            2.250         1.000     1.000
54  24.262 18.493            5.770         2.500     1.000
55  24.086 26.027           -1.941        -7.000     1.000
56  23.049 20.410            2.639         5.000     0.000
57  24.149 19.683            4.466         6.000     0.000
58  24.729 21.090            3.639        -9.000     1.000
59  21.066 22.628           -1.561       -15.000     1.000
60  25.662 21.094            4.568        -2.500     1.000
61  22.598 20.419            2.178         3.000     0.000
62  26.298 20.791            5.507         3.500     1.000
63  24.217 21.659            2.558         1.500     1.000
64  22.764 19.163            3.602         3.000     1.000
65  23.336 21.133            2.204         3.000     0.000
66  24.869 22.074            2.795         3.000     

In [51]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(test['home_bet'], predictions['home_bet'])
matrix = matrix / np.sum(matrix)
matrix

array([[0.2622, 0.2959],
       [0.176 , 0.2659]])

In [52]:
print("Accuracy:", (matrix[0][0] + matrix[1][1]))
print("Precision:", (matrix[0][0] / (matrix[0][0] + matrix[0][1])))
print("Recall:", (matrix[0][0] / (matrix[0][0] + matrix[1][0])))

Accuracy: 0.5280898876404494
Precision: 0.46979865771812074
Recall: 0.5982905982905983


THIS CODE IS COMMENTED OUT AS ITS IT IS SIMPLY RAN TO TEST FOR THE BEST MODEL PARAMETERS. This hypertuning block has been run, determining that a C value of 10 and a gamma value of 0.001 returned the lowest Mean Squared Error(194.681), meaning that these parameters (among those tested) allowed our model to most accurately predict the point differential of the games.

### Fitting, Predicting, and Evaluating

In [41]:
clf = SVR(C=10, gamma=0.001)
clf.fit(trainX, trainY, sample_weight=weights)

predY = clf.predict(testX)
MSE = mean_squared_error(testY,predY)
MSE

194.68108949915006

In [42]:
dfSVM = pd.DataFrame(data={'actual': testY,'pred': predY,'vegas_expected_home_win_margin': test['vegas_expected_home_win_margin']})
dfSVM = dfSVM.apply(calculateClass, axis=1, args=('pred',))
dfSVM

actual   pred  vegas_expected_home_win_margin  class
5057  -7.000  2.693                           3.000  0.000
5058   0.000 -0.855                          -3.000  1.000
5059  -3.000 -1.306                          -2.000  1.000
5060 -30.000  1.739                           5.500  0.000
5061  18.000  3.458                           7.000  0.000
5062 -14.000 -3.409                          -3.500  1.000
5063   6.000  1.146                          -6.000  1.000
5064 -49.000 -2.074                          -7.000  1.000
5065  16.000  5.891                           3.500  1.000
5066  30.000  7.866                           5.500  1.000
5067  -1.000 -0.085                           2.500  0.000
5068   5.000  7.103                          10.000  0.000
5069   1.000  4.285                           9.000  0.000
5070 -14.000  1.112                           1.000  1.000
5071   2.000  5.999                           6.500  0.000
5072   8.000  0.361                          -3.000  1.000
5073  -6.000  4.923                           6.500  0.000
5074   4.000 -1.802                          -1.000  0.000
5075   6.000  5.661                          13.000  0.000
5076 -24.000  0.469                           1.000  0.000
5077  -2.000  4.475                          -3.000  1.000
5078   3.000 -4.268                          -1.000  0.000
5079   5.000  2.459                           3.000  0.000
5080   1.000  4.528                           7.000  0.000
5081  18.000 -3.772                          -2.000  0.000
5082 -43.000 -3.397                         -18.000  1.000
5083 -14.000  0.303                          -1.000  1.000
5084 -18.000 -2.661                          -7.000  1.000
5085  -2.000  5.117                           4.000  1.000
5086  -2.000  4.441                           3.500  1.000
5087 -10.000  0.078                          -6.000  1.000
5088 -20.000  0.976                          -6.500  1.000
5089  13.000  1.744                          -2.000  1.000
5090 -18.000  1.670                           2.500  0.000
5091   4.000  1.030                           6.000  0.000
5092  -7.000 -3.838                          -4.000  1.000
5093  25.000  6.650                          21.500  0.000
5094  11.000  3.323                           7.000  0.000
5095   3.000  0.402                           1.500  0.000
5096   5.000  1.998                           5.000  0.000
5097  -7.000  0.888                          -3.000  1.000
5098  20.000  7.158                           9.000  0.000
5099  16.000 10.737                          21.000  0.000
5100  -3.000  4.779                           4.000  1.000
5101   4.000  0.045                           6.000  0.000
5102  -6.000  0.235                           5.000  0.000
5103  -1.000 -0.353                           5.000  0.000
5104 -16.000  0.810                          -5.000  1.000
5105  -7.000  0.225                           3.500  0.000
5106 -17.000 -0.137                          -5.500  1.000
5107 -14.000  3.523                           3.500  1.000
5108 -15.000  4.471                           7.000  0.000
5109  -6.000 -2.537                          -7.000  1.000
5110  10.000  2.874                           1.000  1.000
5111  -2.000  6.239                           2.500  1.000
5112  -4.000 -3.438                          -7.000  1.000
5113  -6.000  3.671                           5.000  0.000
5114  -7.000  1.725                           6.000  0.000
5115 -15.000  3.357                          -9.000  1.000
5116 -20.000 -3.824                         -15.000  1.000
5117   2.000  5.278                          -2.500  1.000
5118  21.000  1.176                           3.000  0.000
5119  24.000  4.949                           3.500  1.000
5120   1.000  1.017                           1.500  0.000
5121   7.000  4.586                           3.000  1.000
5122  -3.000  2.087                           3.000  0.000
5123 -10.000  2.759                           3.000  

The model is trained and test using the optimal parameters and the sample weights defined above. Afterwards, the actual score differentials, the model's prediction, the vegas spread, and the newly calculated class are shown in a dataframe.

In [43]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(test['class'], dfSVM['class'])
matrix = matrix / np.sum(matrix)
matrix

array([[0.2809, 0.2772],
       [0.191 , 0.2509]])

In [44]:
print("Accuracy:", (matrix[0][0] + matrix[1][1]))
print("Precision:", (matrix[0][0] / (matrix[0][0] + matrix[0][1])))
print("Recall:", (matrix[0][0] / (matrix[0][0] + matrix[1][0])))

Accuracy: 0.5318352059925093
Precision: 0.5033557046979866
Recall: 0.5952380952380952


The predicted class is then compared with the actual class. Again, a '0' or '-' would mean that the model is predicting a bet should be made on the away team to cover the spread. A '1' or a '+' would mean the model is predicting a bet should be made on the home team to cover the spread. The model would be 'correct' if its predicted bet is validated by the point differential being above or below the vegas spread. Accuracy, precision, and recall values are also shown and are unforunately lower than I had hoped for. Since this model did not work out, I will also be attempting an preceptron regressor to ensure my SVM model was not just wrongly prepared.

## Multilayer Perceptron Regressor

### Hypertuning

In [52]:
from sklearn.neural_network import MLPRegressor

trainX = train.drop(columns=['vegas_expected_home_win_margin','home_point_lead','class'])
testX = test.drop(columns=['vegas_expected_home_win_margin','home_point_lead','class'])

alphas = [0.0001,0.001,0.01,0.1]
iters = [100,200,400,600]

results = []

for alpha in alphas:
    for iter in iters:
        regressor = MLPRegressor(hidden_layer_sizes=(600,300),learning_rate_init=alpha,max_iter=iter)
        regressor.fit(trainX,trainY)

        predY = regressor.predict(testX)
        MSE = mean_squared_error(testY,predY)

        results.append([alpha,iter,MSE])

resultsFrame = pd.DataFrame(data=results,columns=['Alpha','Iterations','MSE'])
resultsFrame

Alpha  Iterations     MSE
0   0.000         100 199.421
1   0.000         200 230.789
2   0.000         400 288.758
3   0.000         600 206.537
4   0.001         100 194.506
5   0.001         200 201.525
6   0.001         400 238.462
7   0.001         600 240.624
8   0.010         100 228.684
9   0.010         200 251.171
10  0.010         400 211.596
11  0.010         600 232.172
12  0.100         100 223.362
13  0.100         200 224.321
14  0.100         400 224.150
15  0.100         600 224.894

Like the SVM model, the MLP Regressor will be hypertuned to determine the best parameters. This block of code is also commented out, its compute time also being unnecesssary after the tests. After the hypertuning, it was determined that a learning rate (alpha) of 0.001 and a max iteration count of 100 returned the lowest Mean Squared Error(194.506).

In [53]:
regressor = MLPRegressor(hidden_layer_sizes=(600,300),learning_rate_init=0.001,max_iter=100)
regressor.fit(trainX, trainY)

predY = regressor.predict(testX)
MSE = mean_squared_error(testY,predY)


In [54]:
dfMLR = pd.DataFrame(data={'actual': testY,'pred': predY,'vegas_expected_home_win_margin': test['vegas_expected_home_win_margin']})
dfMLR = dfMLR.apply(calculateClass, axis=1, args=('pred',))
dfMLR

actual   pred  vegas_expected_home_win_margin  class
5057  -7.000  6.997                           3.000  1.000
5058   0.000  1.499                          -3.000  1.000
5059  -3.000  1.982                          -2.000  1.000
5060 -30.000  5.480                           5.500  0.000
5061  18.000  5.966                           7.000  0.000
5062 -14.000  0.423                          -3.500  1.000
5063   6.000  5.856                          -6.000  1.000
5064 -49.000  2.233                          -7.000  1.000
5065  16.000  7.381                           3.500  1.000
5066  30.000 10.407                           5.500  1.000
5067  -1.000  4.540                           2.500  1.000
5068   5.000 10.623                          10.000  1.000
5069   1.000  9.348                           9.000  1.000
5070 -14.000  4.226                           1.000  1.000
5071   2.000  9.470                           6.500  1.000
5072   8.000  4.878                          -3.000  1.000
5073  -6.000  6.754                           6.500  1.000
5074   4.000  2.501                          -1.000  1.000
5075   6.000 11.710                          13.000  0.000
5076 -24.000  1.362                           1.000  1.000
5077  -2.000  7.942                          -3.000  1.000
5078   3.000  3.122                          -1.000  1.000
5079   5.000  6.139                           3.000  1.000
5080   1.000  8.315                           7.000  1.000
5081  18.000  1.477                          -2.000  1.000
5082 -43.000  1.031                         -18.000  1.000
5083 -14.000  6.030                          -1.000  1.000
5084 -18.000  1.997                          -7.000  1.000
5085  -2.000  8.548                           4.000  1.000
5086  -2.000  6.565                           3.500  1.000
5087 -10.000  2.577                          -6.000  1.000
5088 -20.000  3.760                          -6.500  1.000
5089  13.000  6.222                          -2.000  1.000
5090 -18.000  3.490                           2.500  1.000
5091   4.000  6.383                           6.000  1.000
5092  -7.000  2.298                          -4.000  1.000
5093  25.000  8.082                          21.500  0.000
5094  11.000  4.873                           7.000  0.000
5095   3.000  7.111                           1.500  1.000
5096   5.000  4.249                           5.000  0.000
5097  -7.000  6.235                          -3.000  1.000
5098  20.000  8.672                           9.000  0.000
5099  16.000 11.788                          21.000  0.000
5100  -3.000  7.541                           4.000  1.000
5101   4.000  3.730                           6.000  0.000
5102  -6.000  3.815                           5.000  0.000
5103  -1.000  3.204                           5.000  0.000
5104 -16.000  3.829                          -5.000  1.000
5105  -7.000  3.393                           3.500  0.000
5106 -17.000  2.452                          -5.500  1.000
5107 -14.000  6.375                           3.500  1.000
5108 -15.000  8.728                           7.000  1.000
5109  -6.000  3.212                          -7.000  1.000
5110  10.000  6.468                           1.000  1.000
5111  -2.000 10.352                           2.500  1.000
5112  -4.000  0.818                          -7.000  1.000
5113  -6.000  6.592                           5.000  1.000
5114  -7.000  8.052                           6.000  1.000
5115 -15.000  7.654                          -9.000  1.000
5116 -20.000  2.243                         -15.000  1.000
5117   2.000  8.363                          -2.500  1.000
5118  21.000  5.736                           3.000  1.000
5119  24.000 10.249                           3.500  1.000
5120   1.000  6.371                           1.500  1.000
5121   7.000  6.952                           3.000  1.000
5122  -3.000  4.827                           3.000  1.000
5123 -10.000  6.527                           3.000  

In [55]:
matrix = confusion_matrix(test['class'], dfMLR['class'])
matrix = matrix / np.sum(matrix)
matrix

array([[0.1199, 0.4382],
       [0.0787, 0.3633]])

In [56]:
print("Accuracy:", (matrix[0][0] + matrix[1][1]))
print("Precision:", (matrix[0][0] / (matrix[0][0] + matrix[0][1])))
print("Recall:", (matrix[0][0] / (matrix[0][0] + matrix[1][0])))

Accuracy: 0.48314606741573035
Precision: 0.21476510067114093
Recall: 0.6037735849056605


Like the SVM model, the predicted class is then compared with the actual class. Our evalutation metrics for the derived classification are once agan lower than expected.

In [57]:
MSE = mean_squared_error(testY,test['vegas_expected_home_win_margin'])
MSE

170.9765917602996

Here, the MSE between the actual point differntials and the vegas spread is shown. Since this is esssentially another regression prediction, it can be used as a good comparison to our models' performances. Unfortunately, at this time I was not able to outperform the Vegas models. In fact, my methods underperformed greatly when compared to these models (delta of 24). The gambling industry, most likely, has much better resources, datasets, and modeling procedures than I was able to apply for the scope of this project. In the future, I would like to continue to tweak this project to come closer to outperforming the Vegas models.

## Conclusion and Wrap Up

The project, as a whole, was a lot of fun to map out and execute. It was very interesting to try and think about the game of football as a math equation; with hundreds of variables of differing influence. While the results founded were not exactly what I was hoping for, I do plan on continuing on tweaking the project, most likely in the form of gathering more relevant data and formatting it in a better format. I would also consider using Principle Component Analysis on the data since currently it has far too many features. This resulted in hypertuning my models to run for far too long, wasting a lot of time while I waited for them to execute. Overall, this project taught me the many complexities and hardships of data preperation and modeling. Despite it being very challenging, it also was very engaging and would be happy to carry on doing this in the future.

### Custom Programming and Bibliography

* http://www.jt-sw.com/football/pro/rosters.nsf used to get all team rosters since the year 2000
* https://www.kaggle.com/tobycrabtree/nfl-scores-and-betting-data#spreadspoke_scores.csv used to get game information
* Custom programming work done includes most of the preprocessing code. This includes the web scraping block, calculating wins and losses for each game, adding the players to the game data entries, and calculating the class.